data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ParserTest.java

```diff
@@ -0,0 +1,142 @@
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.tree.ParseTreeWalker;
+import org.junit.jupiter.api.Test;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionLexer;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
+import org.opensearch.dataprepper.expression.util.ErrorListener;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class ParserTest {
+
+    private ErrorListener errorListener;
+
+    private void parseExpression(final String expression) {
+        errorListener = new ErrorListener();
+
+        final CodePointCharStream stream = CharStreams.fromString(expression);
+        final DataPrepperExpressionLexer lexer = new DataPrepperExpressionLexer(stream);
+        lexer.addErrorListener(errorListener);
+
+        final CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+        final DataPrepperExpressionParser parser = new DataPrepperExpressionParser(tokenStream);
+        parser.addErrorListener(errorListener);
+
+        final ParserRuleContext context = parser.expression();
+
+        final ParseTreeWalker walker = new ParseTreeWalker();
+        walker.walk(errorListener, context);
+    }
+
+    private void assertThatIsValid(final String expression) {
+        parseExpression(expression);
+        assertThat(
+                "\"" + expression + "\" should not have parsing errors",
+                errorListener.isErrorFound(),
+                is(false)
+        );
+    }
+
+    private void assertThatHasParseError(final String expression) {
+        parseExpression(expression);
+        assertThat(
+                "\"" + expression + "\" should have parsing errors",
+                errorListener.isErrorFound(),
+                is(true)
+        );
+    }
+
+    @Test
+    void testGivenOperandIsNotStringWhenEvaluateThenErrorPresent() {
+        assertThatHasParseError("(true) =~ \"Hello?\"");
+    }
+
+    @Test
+    void testInvalidSetItem() {
+        assertThatHasParseError("{true or false}");
+        assertThatHasParseError("{(5)}");
+        assertThatHasParseError("{5 == 5}");
+        assertThatHasParseError("{/node =~ \"[A-Z]*\"}");
+    }
+
+    @Test
+    void testInvalidSetOperator() {
+        assertThatHasParseError("1 in 1");
+    }
+
+    @Test
+    void testInvalidRegexOperand() {
+        assertThatHasParseError("\"foo\"=~3.14");
+    }
+
+    @Test
+    void testInvalidFloatParsingRules() {
+        assertThatHasParseError("0.");
+        assertThatHasParseError("00.");
+        assertThatHasParseError(".00");
+        assertThatHasParseError(".10");
+        assertThatHasParseError("1.10");
+    }
+
+    @Test
+    void testInvalidNumberOfOperands() {
+        assertThatHasParseError("5==");
+        assertThatHasParseError("==");
+        assertThatHasParseError("not");
+    }
+
+    @Test
+    void testValidOperatorsRequireSpace() {
```

Comment:
These tests could be made into `@ParameterizedTest`s using `@ValueSource`. But, they should still be fine this way.

```
@ParameterizedTest
@ValueSource(strings = {
  "/status in {200}",
  "/a==(/b==200)",
  ...
})
void testValidOperatorsRequireSpace(String expression) {
  assertThatIsValid(expression);
}
```

Reply:
Great idea on the `@ParameterizedTest`. It gives the benefits of `assertAll(...)` with more descriptive error messages!